dialogue-core/src/main/java/com/palantir/dialogue/core/StickyEndpointChannels.java

```diff
@@ -0,0 +1,163 @@
+        this.tracker = new BalancedScoreTracker(
+                params.channels.size(),
+                params.random,
+                params.ticker,
+                Preconditions.checkNotNull(params.taggedMetrics, "taggedMetricRegistry"),
+                Preconditions.checkNotNull(params.channelName, "channelName"));
+    }
+
+    /**
+     * Returns the current 'best' channel based on its Balanced score. If the underlying host is completely offline,
+     * callers are responsible for getting a new sticky channel by calling this method again, relying on the
+     * {@link BalancedScoreTracker} to avoid returning the same broken channel again.
+     */
+    public Channel getStickyChannel() {
+        return new Sticky(channels, tracker);
+    }
+
+    @ThreadSafe
+    private static final class Sticky implements EndpointChannelFactory, Channel {
+
+        private final ImmutableList<? extends EndpointChannelFactory> channels;
+        private final Supplier<ChannelScoreInfo> getSingleBestChannel;
+
+        public Sticky(ImmutableList<? extends EndpointChannelFactory> channels, BalancedScoreTracker tracker) {
+            this.channels = channels;
+            this.getSingleBestChannel = Suppliers.memoize(tracker::getSingleBestChannelByScore);
+        }
+
+        @Override
+        public EndpointChannel endpoint(Endpoint endpoint) {
+            ChannelScoreInfo channelScoreInfo = getSingleBestChannel.get();
+            int hostIndex = channelScoreInfo.channelIndex();
+            EndpointChannel delegate = channels.get(hostIndex).endpoint(endpoint);
+            return new ScoreTrackingEndpointChannel(delegate, channelScoreInfo);
+        }
+
+        /**
+         * .
+         * @deprecated prefer {@link #endpoint}, as it allows binding work upfront
+         */
+        @Deprecated
+        @Override
+        public ListenableFuture<Response> execute(Endpoint _endpoint, Request _request) {
+            // TODO(dfox): could we delete this entirely?
+            throw new UnsupportedOperationException("Not implemented");
+        }
+    }
+
+    private static final class ScoreTrackingEndpointChannel implements EndpointChannel {
+        private final EndpointChannel delegate;
+        private final ChannelScoreInfo tracker;
+
+        ScoreTrackingEndpointChannel(EndpointChannel delegate, ChannelScoreInfo tracker) {
+            this.delegate = delegate;
+            this.tracker = tracker;
+        }
+
+        @Override
+        public ListenableFuture<Response> execute(Request request) {
+            tracker.startRequest();
+            ListenableFuture<Response> future = delegate.execute(request);
+            tracker.observability().markRequestMade();
+            DialogueFutures.addDirectCallback(future, tracker);
+            return future;
+        }
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
```

Comment:
nit: private constructor. I get snagged by this all the time, I'll try to carve out time to write an error-prone rule at some point.